📖 document automatic certificate renewal support in KCP (#7268)
kubernetes-sigs/cluster-api: docs/book/src/tasks/certs/auto-rotate-certificates-in-kcp.md

## Automatically rotating certificates using Kubeadm Control Plane provider

> When using Kubeadm Control Plane provider (KCP) it is possible to configure automatic certificate rotations. KCP does this by triggering a rollout when the certificates on the machines are about to expire.

Reviewer: Let's add a note explaining that this is automatically enabled only for KCP-managed machines created with CAPI >= v1.3; also add a paragraph about how to set this for existing machines (ssh into the machine, run kubeadm command to get certificate expiration date, set the annotation on the machine) and reference it from here.

### Configuring Machine Rollout

To configure a rollout on the KCP machines you need to set `.rolloutBefore.certificatesExpiryDays` (minimum of 7 days).  

Example: 
```yaml
apiVersion: controlplane.cluster.x-k8s.io/v1beta1
kind: KubeadmControlPlane
metadata:
  name: example-control-plane
spec:
  rolloutBefore:
    certificatesExpiryDays: 21 # trigger a rollout if certificates expire within 21 days
  kubeadmConfigSpec:
    clusterConfiguration:
      ...
    initConfiguration:
      ...
    joinConfiguration:
      ...
  machineTemplate:
    infrastructureRef:
      ...
  replicas: 1
  version: v1.23.3
``` 

> TODO: Add a few lines on recommendations to set the `certificatesExpiryDays` to a big value so that all the machines will have time to completely rollout.

Reviewer: s/all the machines will have time to completely rollout/rollout is completed well in advance before expiration.

> ### Calculating Certificate Expiry

Reviewer: ```suggestion
### Triggering Machine Rollout for certificate expiry
```

KCP uses the value in the corresponding Control Plane machine's `Machine.Status.CertificatesExpiryDate` to check if a machine's certificates are going to expire and if it needs to be rolled out.  

`Machine.Status.CertificatesExpiryDate` get its value from one of the following 2 places:

> * `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Machine object

Reviewer: ```suggestion
* `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Machine object; this annotation is not applied by default and it can be set by users to opt-in in this feature for existing control plane machines (see instructions below)
```

> * `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Bootstrap Config object of the corresponding object

Reviewer: ```suggestion
* `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Bootstrap Config object referenced by the machine. this value is automatically set for KCP-managed machines created with CAPI >= v1.3
```
And thus I will drop "Certificate Expiry from Kubeadm Bootstrap Config" paragraph except for the aside which makes sense to have.

The annotation value is a [RFC3339] format timestamp. The annotation value on the machine object, if provided, will take precedence.  

#### Certificate Expiry from Kubeadm Bootstrap Config

When KubeadmConfig creates the bootstrap data secret for the first time it captures 1 year from the reconciliation time and store it as the `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the KubeadmConfig object. We chose 1 year from the reconciliation time because kubeadm is hard-coded to generate machine certificates that are valid for one year.

<aside class="note warning">

<h1> Approximate Certificate Expiry Time </h1>

> The time captured in the KubeadmConfig annotation is an approximate time at which the machine certificates will expire. Since the Kubeadm data secret is generated before is is used by kubeadm to generate the certificates the time captured in the annotation will be a little early than the actual certificate expiry time.

Reviewer: ```suggestion
The time captured in the KubeadmConfig annotation is an approximate time at which the machine certificates will expire in 1 year time; the time captured in the annotation will be a little earlier than the actual certificate expiry time.
```

> Since the difference in this time will be very small compared to the lifetime of the certificate it should not affect the certificate renewal rollout in any significant way.

Reviewer: ```suggestion
```

## Automatically rotating certificates using Kubeadm Control Plane provider

> When using Kubeadm Control Plane provider (KCP) it is possible to configure automatic certificate rotations. KCP does this by triggering a rollout when the certificates on the machines are about to expire.

Reviewer 2: ```suggestion
When using Kubeadm Control Plane provider (KCP) it is possible to configure automatic certificate rotations. KCP does this by triggering a rollout when the certificates on the control plane machines are about to expire.
```
I know it can be inferred but maybe good to emphasize

> Rollout for certificate expiry is automatically enabled for all new clusters created using ClusterAPI v1.3 and above. To enable it for existing clusters see [Enabling Certificate Expiry Rollout for Existing Clusters](#enabling-certificate-expiry-rollout-for-existing-clusters)

Author: >It happens on a per-machine basis, correct? Otherwise folks might think it's only on new clusters
Yes, it is on a machine basis.

>Not sure if we should say automatically enabled given that it's only enabled if certificatesExpiryDays is set
I re-worede the statement a little to adjust for this. WDYT?

> Rollout for certificate expiry is automatically enabled for all new clusters created using ClusterAPI v1.3 and above. To enable it for existing clusters see [Enabling Certificate Expiry Rollout for Existing Clusters](#enabling-certificate-expiry-rollout-for-existing-clusters)

Reviewer 2: Sounds good!

> If configured, rollout for certificate expiry is enabled for all new machines created using ClusterAPI v1.3 and above. To enable it for existing machines see [Enabling Certificate Expiry Rollout for Existing Machines](#enabling-certificate-expiry-rollout-for-existing-clusters)(#enabling-certificate-expiry-rollout-for-existing-clusters)

Reviewer: ```suggestion
If configured, rollout for certificate expiry is enabled for all new control plane machines created using ClusterAPI v1.3 and above. To enable it for existing control plane machines see [Enabling Certificate Expiry Rollout for Existing Machines](#enabling-certificate-expiry-rollout-for-existing-clusters)(#enabling-certificate-expiry-rollout-for-existing-clusters)
```
or KCP machines

Reviewer 2: + please update the link in this line as the headline was changed (machines instead of clusters)

kind: KubeadmControlPlane
metadata:
  name: example-control-plane
spec:
  rolloutBefore:
    certificatesExpiryDays: 21 # trigger a rollout if certificates expire within 21 days
  kubeadmConfigSpec:
    clusterConfiguration:
      ...
    initConfiguration:
      ...
    joinConfiguration:
      ...
  machineTemplate:
    infrastructureRef:
      ...
  replicas: 1
  version: v1.23.3
``` 

It is strongly recommended to set the `certificatesExpiryDays` to a large enough value so that all the machines will have time to complete rollout well in advance before the certificates expire.

### Triggering Machine Rollout for Certificate Expiry

KCP uses the value in the corresponding Control Plane machine's `Machine.Status.CertificatesExpiryDate` to check if a machine's certificates are going to expire and if it needs to be rolled out.  

`Machine.Status.CertificatesExpiryDate` gets its value from one of the following 2 places:

* `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Machine object. This annotation is not applied by default and it can be set by users to opt-in in this feature for existing control plane machines (see instructions below).

> * `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Bootstrap Config object referenced by the machine. This value is automatically set for KCP-managed machines created with CAPI >= v1.3.

Reviewer 2: ```suggestion
* `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Bootstrap Config object referenced by the machine. This value is automatically set for machines bootstrapped with CABPK with CAPI >= v1.3.
```

Reviewer: if we use bootstrapped with CABPK we should clarify that this apply only to CP machines. (KCP-managed implicitly means both)

Reviewer 2: Hm KCP is only compatible with CABPK, right? Then let's keep the current wording for simplicity

The annotation value is a [RFC3339] format timestamp. The annotation value on the machine object, if provided, will take precedence.  

<aside class="note warning">

<h1> Approximate Certificate Expiry Time </h1>

The time captured in the Bootstrap Config annotation is an approximate time at which the machine certificates will expire (1 year from creation). The time captured in the annotation will be a little earlier than the actual certificate expiry time.

</aside>

> ### Enabling Certificate Expiry Rollout for Existing Clusters

Reviewer 2: ```suggestion
### Enabling Certificate Expiry Rollout for Existing Machines
```

Reviewer 2: same below

For rollout for certificate expiry to work on existing machines we need to update existing control plane machines with the current certificate expiry time. To do this do the following for each of the control plane machines:

* SSH into the control plane node and get the current certificate expiry time using `kubeadm certs check-expiration`. From the output note the certificate expiry time of the `apiserver` certificate.

> * Add the certificate expiry time as the `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the control plane machine object. The value should be a [RFC3339] format timestamp.

Reviewer 3: Can't Cluster API / KCP introspect these values by running an exec pod temporarily on the node?

Reviewer: We can open an issue to discuss this idea, It is doable but at first sight, this isn't a small amount of work (mostly because most probably this requires a new docker image + all the corresponding release machinery); there are also security concerns given that the pod need access to sensitive material

Author: Issue: https://github.com/kubernetes-sigs/cluster-api/issues/7342

  replicas: 1
  version: v1.23.3
``` 

It is strongly recommended to set the `certificatesExpiryDays` to a large enough value so that all the machines will have time to complete rollout well in advance before the certificates expire.

### Triggering Machine Rollout for Certificate Expiry

KCP uses the value in the corresponding Control Plane machine's `Machine.Status.CertificatesExpiryDate` to check if a machine's certificates are going to expire and if it needs to be rolled out.  

`Machine.Status.CertificatesExpiryDate` gets its value from one of the following 2 places:

* `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Machine object. This annotation is not applied by default and it can be set by users to manually override the certificate expiry information.
* `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Bootstrap Config object referenced by the machine. This value is automatically set for machines bootstrapped with CABPK that are owned by the KCP resource.

The annotation value is a [RFC3339] format timestamp. The annotation value on the machine object, if provided, will take precedence.  

<aside class="note warning">

> <h1> Approximate Certificate Expiry Time </h1>

Reviewer 2: ```suggestion
<h1>Certificate Expiry Time </h1>
```
maybe, up to you

Reviewer: I think we can drop this aside, we are reading the actual expiry date from the API server,

Reviewer 2: Agree, that's what I suggested here: https://github.com/kubernetes-sigs/cluster-api/pull/7268#discussion_r1012867021

The time captured in the Bootstrap Config annotation is an approximate time at which the machine certificates will expire (1 year from creation). The time captured in the annotation will be a little earlier than the actual certificate expiry time.

It is assumed that all the certificates has the same expiration time. If not, it is assumed that the kube-apiserver certificate expires before other certificates.

</aside>

> 

Reviewer 2: Let's add a new warning about "Manual certificate rotation"

Roughly:

```markdown
<h1>Manual certificate rotation</h1>

If certificates on control plane nodes are rotated manually (e.g. via `kubeadm certs renew`), please be aware that the 
rotation is only complete after all components including the kube-apiserver are using the new certificates. Thus, 
kube-apiserver, kube-controller-manager, kube-scheduler and etcd have to be restarted after certificate renewal.
To allow KCP to re-discover the expiry date please remove the `machine.cluster.x-k8s.io/certificates-expiry` annotation
from the KubeadmConfig corresponding to the current machine.

</aside>
```

> ## Automatically rotating certificates using Kubeadm Control Plane provider

Reviewer 2: Just a blocker comment with notes for now. I'll update depending on the outcome of some discussions and the other PR:
* We should document that we assume all certificates are generated with the same expiry
* [TBD] We should document that a manual certificate rotation is only complete after the kube-apiserver is serving the new certs

(only applies if the new PR is merged)

Author: >We should document that a manual certificate rotation is only complete after the kube-apiserver is serving the new certs and the expiry annotation has been removed from KubeadmConfig

The annotation will be added back on the new KubeadmConfig by KCP, isnt it?

Reviewer 2: Added a suggestion below for a warning

`Machine.Status.CertificatesExpiryDate` gets its value from one of the following 2 places:

* `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Machine object. This annotation is not applied by default and it can be set by users to opt-in in this feature for existing control plane machines (see instructions below).
* `machine.cluster.x-k8s.io/certificates-expiry` annotation value on the Bootstrap Config object referenced by the machine. This value is automatically set for machines bootstrapped with CABPK with CAPI >= v1.3.

The annotation value is a [RFC3339] format timestamp. The annotation value on the machine object, if provided, will take precedence.  

<aside class="note warning">

<h1> Approximate Certificate Expiry Time </h1>

The time captured in the Bootstrap Config annotation is an approximate time at which the machine certificates will expire (1 year from creation). The time captured in the annotation will be a little earlier than the actual certificate expiry time.

</aside>

<aside class="note warning">

> <h1> Deleting the machine.cluster.x-k8s.io/certificates-expiry annotation </h1>

Author: This is not longer true, isn't it?
If the annotation is deleted then KCP reconciler will add the annotation back on the KubeadmConfig and the certificate rollout should keep working, correct?

We can safely drop this warning, correct?

Reviewer 2: Yes, correct

If the annotation is delete from the object, the certificate expiry information will be cleared form the Machine's status leading to certificate renewal being effectively disabled. It is recommended to be highly cautions when deleting this annotation from the object.

</aside>

> ### Enabling Certificate Expiry Rollout for Existing Control Plane Machines

Author: With the auto-discovery in KCP we can drop this section, right?

Reviewer 2: Yes!
